docs/TeslaSleepConfiguration.md

```diff
@@ -0,0 +1,118 @@
+# Tesla Sleep Configuration
+
+## Introduction
+
+Tesla vehicles have a sleep mode which allows the vehicle to conserve battery power when not actively operating. After a period of inactivity, the vehicle will power down systems which are not necessary during times of inactivity. During these periods, your vehicle will only awaken if:
+
+  * You approach the vehicle within range using the Android or iPhone application over bluetooth.
+
+One challenge of Tesla API-based Data Loggers is that the continual polling of vehicles via the API results in the vehicle waking up or failing to go into sleep mode. To avoid these scenarios, TeslaMate uses two configuration parameters to define when a vehicle will start to sleep due to inactivity.
+
+## Time to Try Sleeping
+
+Following activity such as driving or charging, TeslaMate will continue to actively poll a vehicle for a configurable amount of time (by default **15 Minutes**) in order to detect whether the vehicle will continue to be in active use (for example, whether the vehicle has stopped temporarily to pick up a passenger or to visit a shop) or whether it will return to idle state.
+
+  * If this value is set too low, the car may attempt to sleep prior to actually being idle. In this case, you may miss subsequent parts of a trip, for example if you had stopped for 20 minutes to pick up some items but the idle time was set to 15 minutes, the car would attempt to sleep.
+  * If this value is set too high, the time after a drive during which the car is continually actively polled and is unable to go to sleep will be longer than necessary, consuming a greater amount of battery power.
+
+## Attempt to Sleep
+
+Once the Time to Try Sleeping idle timer is reached, TeslaMate will try to allow the car to go to sleep. It does this by halting all polling for the configured time period, which is **21 Minutes** by default. 
+
+  * You will know that the vehicle is in this state when the status in TeslaMate is *falling asleep for x minutes*.
+  * At the end of this polling window, TeslaMate will poll the vehicle status. If the vehicle status is offline, this indicates that the sleep process succeeded and the car is no longer directly queryable.
+  * Polling will continue once per configured interval (*by default, 21 minutes*), meaning TeslaMate's resolution of polling interval for vehicles which have gone into sleep mode is 21 minutes - unless otherwise notified, TeslaMate will not know that the vehicle is awake for a total of 21 minutes (*by default*).
```

Comment:
One minor correction: the resolution of the polling interval for vehicles which have gone into sleep mode is __1 minute__ not 21. That means if the car is in sleep mode TeslaMate will notice when it wakes up with a short delay of up to 60 seconds.
The worst-case, i.e.  missing 21 minutes of a drive session. only occurs if the vehicle wakes up from sleep mode right after TeslaMate halts polling.

Reply:
Thanks for the clarification Adrian! One of the reasons I wanted to go into this level of detail in the document was because I wasn't 100% sure that I understood the mechanism and wanted to break it down to do so.

Just so that I can better elaborate the process in this document, would this be accurate from a high level perspective:

- If TeslaMate polls the vehicle and it is one of (eg): driving, charging - it is online and will be polled regularly (possible to get the exact resolution of polling?)
- If the vehicle is online but not otherwise active, it is polled once per minute until we reach the idle period (which is 15 minutes by default)
- *My assumption at this point was that it would then revert to a poll once per (configured sleep period) to allow the car to attempt to sleep, sounds like however it is still polling every minute during this period?*
- Once the car is confirmed offline, it will be polled once per configured sleep polling period (once per 21 minutes by default). If it remains offline, this polling remains at this resolution, however as soon as it is polled as online, the polling interval reverts to one per minute

Also, just so I fully understand the resume/suspend mechanism:

- Resume would cause TeslaMate to revert to once per minute polling (for a given period of time?) even if the car currently shows offline, to allow it to pick up an imminent online event.
- Suspend would cause TeslaMate to revert to the offline polling schedule of (by default 21 minutes)?

Reply:
Yeah, sure. I think it's worth to explain it in detail!

> If TeslaMate polls the vehicle and it is one of (eg): driving, charging - it is online and will be polled regularly (possible to get the exact resolution of polling?)

Exactly. When driving it polls every 2.5 seconds. I'd consider this an implementation detail and it might change slightly in the future. But in general the vehicle is being polled  _every few seconds_. 

When charging, the polling interval depends on the charger power: The higher the power, the higher the polling frequency. The resolution can range from a few seconds to 30 seconds.

>  If the vehicle is online but not otherwise active, it is polled once per minute until we reach the idle period (which is 15 minutes by default)

If it is online but not otherwise active, the polling frequency depends on a few variables e.g. 

- is a driver still present
- is the vehicle unlocked
- is sentry mode active
- ...

In all cases the vehicle is being polled regularly every 10 to 30 seconds. This process takes 15 minutes by default or until we send the suspend command.

> My assumption at this point was that it would then revert to a poll once per (configured sleep period) to allow the car to attempt to sleep, sounds like however it is still polling every minute during this period?

I was a little ambiguous here. After the idle period of 15 minutes, polling is halted (for 21 minutes by default). AFAIK the vehicle can only go into sleep mode if it's not being polled. This is the most "critical" part which can potentially cause a data gap if a drive session is started during that period because TeslaMate simply cannot know. Basically, this period is the reason why there is the `resume` command. At this point, we have to instruct TeslaMate to continue polling if we start to drive (or charge).

> Once the car is confirmed offline, it will be polled once per configured sleep polling period (once per 21 minutes by default). If it remains offline, this polling remains at this resolution, however as soon as it is polled as online, the polling interval reverts to one per minute

Once it is offline (or asleep) it will be polled every minute. You may be wondering now why the heck the car continues to sleep although we start polling again :-) 

The reason is that there are two different kinds of API requests to fetch vehicle data. The one that gets us all data (position, speed, charger power etc.) is only possible if the car is online. This kind of request prevents the vehicle from falling asleep. 
Then there's another one that can be made even though the car is offline/asleep. This request returns not that much useful information but it contains the vehicle status. This makes it possible to know when the car is waking up.

> Resume would cause TeslaMate to revert to once per minute polling (for a given period of time?) even if the car currently shows offline, to allow it to pick up an imminent online event.
    Suspend would cause TeslaMate to revert to the offline polling schedule of (by default 21 minutes)?

Like mentioned above `resume` "reverts" the halted polling or in other words it instructs TeslaMate to start polling again, if it was halted earlier.

Now, to make it even more complex, `resume` was given a second task:  if the vehicle is offline/asleep (= polling resolution of 1 minute) and we send the `resume` command it will increase the polling frequency so that TM can almost immediately know when the vehicle wakes up. The use case is that my smartphone is often already connected to the car (which as a result is online) but TM would only notice this with a short delay. Sending `resume` will speed up the process.

Reply:
That's 100% clear, thanks Adrian. I'll do some work to adjust the wording around this but I think with that we'll all understand the mechanism and the reasons for the resume/suspend interface